coderd/workspaceupdates.go

```diff
@@ -0,0 +1,299 @@
+package coderd
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/pubsub"
+	"github.com/coder/coder/v2/coderd/util/slice"
+	"github.com/coder/coder/v2/coderd/wspubsub"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/tailnet"
+	"github.com/coder/coder/v2/tailnet/proto"
+)
+
+type workspacesByID = map[uuid.UUID]ownedWorkspace
+
+type ownedWorkspace struct {
+	WorkspaceName string
+	Status        proto.Workspace_Status
+	Agents        []database.AgentIDNamePair
+}
+
+// Equal does not compare agents
+func (w ownedWorkspace) Equal(other ownedWorkspace) bool {
+	return w.WorkspaceName == other.WorkspaceName &&
+		w.Status == other.Status
+}
+
+type sub struct {
+	mu     sync.RWMutex
+	userID uuid.UUID
+	tx     chan<- *proto.WorkspaceUpdate
+	prev   workspacesByID
+
+	db     UpdateQuerier
+	ps     pubsub.Pubsub
+	logger slog.Logger
+
+	cancelFn func()
+}
+
+func (s *sub) ownsAgent(agentID uuid.UUID) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	for _, agent := range s.prev {
+		for _, a := range agent.Agents {
+			if a.ID == agentID {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (s *sub) handleEvent(ctx context.Context, event wspubsub.WorkspaceEvent) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	switch event.Kind {
+	case wspubsub.WorkspaceEventKindStateChange:
+	case wspubsub.WorkspaceEventKindAgentConnectionUpdate:
+	case wspubsub.WorkspaceEventKindAgentTimeout:
+	case wspubsub.WorkspaceEventKindAgentLifecycleUpdate:
```

Comment:
We might be able to get away with only querying on `ConnectionUpdate` and `StateChange`